### PR TITLE
RPC

### PR DESCRIPTION
**Pull request by devsnek**
**Please describe the changes this PR makes and why it should be merged:**
Adds support for discord's rpc over IPC ad WS

**Semantic versioning classification:**  
- [ ] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by devsnek** [comment]
this should not be merged until #1775

File: package.json [hunk 1]
```diff
@@ -49,11 +49,12 @@
   },
   "devDependencies": {
     "@types/node": "^8.0.0",
-    "discord.js-docgen": "hydrabolt/discord.js-docgen",
+    "discord.js-docgen": "github:hydrabolt/discord.js-docgen",
```
Comment (iCrawl) [inline package.json]: Nuh. Remove the `github:`

File: package.json [hunk 2]
```diff
@@ -49,11 +49,12 @@
...
     "eslint": "^4.0.0",
+    "jsdoc-strip-async-await": "^0.1.0",
     "parallel-webpack": "^2.0.0",
     "uglifyjs-webpack-plugin": "^1.0.0-beta.1",
-    "webpack": "^3.0.0"
+    "webpack": "^3.5.4"
```
Comment (iCrawl) [inline package.json]: Stick to ^3.0.0

File: src/BaseClient.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
+const os = require('os');
+const EventEmitter = require('events');
+const RESTManager = require('./rest/RESTManager');
+const ClientDataResolver = require('./client/ClientDataResolver');
+const Util = require('./util/Util');
+const Constants = require('./util/Constants');
+
+/**
+ * The base class for all clients
```
Comment (iCrawl) [inline src/BaseClient.js]: Class descriptions have a dot at the end.

File: src/BaseClient.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+   * API shortcut
+   * @type {Object}
+   * @private
+   */
+  get api() {
+    return this.rest.api;
+  }
+
+  /**
+   * Destroys all assets used by the base client
```
Comment (iCrawl) [inline src/BaseClient.js]: Methods have a dot at the end.

File: src/util/Constants.js [hunk 5]
```diff
@@ -696,3 +703,89 @@ exports.APIErrors = {
...
+  GET_GUILD_TIMED_OUT: 5002,
+  SELECT_VOICE_FORCE_REQUIRED: 5003,
+  CAPTURE_SHORTCUT_ALREADY_LISTENING: 5004,
+};
+
+exports.RPCCloseCodes = {
+  CLOSE_NORMAL: 1000,
+  CLOSE_UNSUPPORTED: 1003,
+  CLOSE_ABNORMAL: 1006,
+  INVALID_CLIENTID: 4e3,
```
Comment (hydrabolt) [inline src/util/Constants.js]: consistency pls???

File: src/util/Constants.js [hunk 6]
```diff
@@ -571,6 +571,13 @@ exports.UserFlags = {
   HYPESQUAD: 1 << 2,
 };
 
+exports.KeyTypes = {
```
Comment (hydrabolt) [inline src/util/Constants.js]: Needs to be documented

File: package.json [hunk 7]
```diff
@@ -54,7 +54,7 @@
     "jsdoc-strip-async-await": "^0.1.0",
     "parallel-webpack": "^2.0.0",
     "uglifyjs-webpack-plugin": "^1.0.0-beta.1",
-    "webpack": "^3.0.0"
+    "webpack": "^3.5.4"
```
Comment (iCrawl) [inline package.json]: Nuh! ^3.0.0, we only lock versions at release for the purpose of keeping them up2date to see if they break something. No reason to always update the numbers.

**Pull request closed without merging**